core/trino-main/src/main/java/io/trino/simd/BlockEncodingSimdSupport.java

```diff
@@ -0,0 +1,117 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.simd;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import io.trino.FeaturesConfig;
+import io.trino.util.MachineInfo;
+import jdk.incubator.vector.ByteVector;
+import jdk.incubator.vector.IntVector;
+import jdk.incubator.vector.LongVector;
+import jdk.incubator.vector.ShortVector;
+import oshi.hardware.CentralProcessor.ProcessorIdentifier;
+
+import java.util.EnumSet;
+import java.util.Set;
+
+import static io.trino.FeaturesConfig.BlockSerdeVectorizedNullSuppressionStrategy.AUTO;
+import static io.trino.util.MachineInfo.readCpuFlags;
+import static java.util.Locale.ENGLISH;
+
+/*
+We need to specifically detect AVX512F (for VPCOMPRESSD / VPCOMPRESSQ instruction support for int and long types) and
+AVX512VBMI2 (for VPCOMPRESSB / VPCOMPRESSW instruction support over byte and short types). Because we would like to check
+whether Vector<T>#compress(VectorMask<T>) is supported natively in hardware or emulated by the JVM - because the emulated
+support is so much slower than the simple scalar code that exists, but since we don't have the ability to detect that
+directly from the JDK vector API we have to assume that native support exists whenever the CPU advertises it.
+ */
+@Singleton
+public final class BlockEncodingSimdSupport
+{
+    public record SimdSupport(
+            boolean expandAndCompressByte,
+            boolean expandAndCompressShort,
+            boolean expandAndCompressInt,
+            boolean expandAndCompressLong)
+    {
+        public static final SimdSupport NONE = new SimdSupport(false, false, false, false);
+        public static final SimdSupport ALL = new SimdSupport(true, true, true, true);
+    }
+
+    public static final int MINIMUM_SIMD_LENGTH = 512;
+    private final SimdSupport simdSupport;
+    private static final SimdSupport AUTO_DETECTED_SUPPORT = detectSimd();
+
+    @Inject
+    public BlockEncodingSimdSupport(
+            FeaturesConfig featuresConfig)
+    {
+        this(featuresConfig.getBlockSerdeVectorizedNullSuppressionStrategy().equals(AUTO));
+    }
+
+    public BlockEncodingSimdSupport(
+            boolean enableAutoDetectedSimdSupport)
+    {
+        if (enableAutoDetectedSimdSupport) {
+            simdSupport = AUTO_DETECTED_SUPPORT;
+        }
+        else {
+            simdSupport = SimdSupport.NONE;
+        }
+    }
+
+    private static SimdSupport detectSimd()
+    {
+        ProcessorIdentifier id = MachineInfo.getProcessorInfo();
+
+        String vendor = id.getVendor().toLowerCase(ENGLISH);
+
+        if (vendor.contains("intel") || vendor.contains("amd")) {
```

Comment:
Lack of ARM support is a bummer. Graviton machines do so much better in the benchmarks that its heavily favored in production deployments at the moment and an incremental gain from better vectorization on intel wouldn't move the needle in the other direction.